cloud/scope/clients.go

```diff
@@ -0,0 +1,169 @@
+/*
+Copyright (c) 2022, Oracle and/or its affiliates.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scope
+
+import (
+	"sync"
+
+	"github.com/go-logr/logr"
+	"github.com/oracle/cluster-api-provider-oci/cloud/config"
+	"github.com/oracle/cluster-api-provider-oci/cloud/ociutil"
+	"github.com/oracle/cluster-api-provider-oci/cloud/services/compute"
+	"github.com/oracle/cluster-api-provider-oci/cloud/services/vcn"
+	"github.com/oracle/oci-go-sdk/v63/common"
+	"github.com/oracle/oci-go-sdk/v63/core"
+	"github.com/oracle/oci-go-sdk/v63/identity"
+	"github.com/oracle/oci-go-sdk/v63/networkloadbalancer"
+	"github.com/pkg/errors"
+	"k8s.io/klog/v2/klogr"
+)
+
+// OCIClients is the struct of all the needed OCI clients
+type OCIClients struct {
+	ComputeClient      compute.ComputeClient
+	VCNClient          vcn.Client
+	LoadBalancerClient *networkloadbalancer.NetworkLoadBalancerClient
+	IdentityClient     *identity.IdentityClient
+}
+
+// ClientProvider defines the regional clients
+type ClientProvider struct {
+	*logr.Logger
+	ociClients            map[string]OCIClients
+	ociClientsLock        *sync.RWMutex
+	ociAuthConfigProvider common.ConfigurationProvider
+	authConfig            *config.AuthConfig
+}
+
+// NewClientProvider builds the ClientProvider with a client for the given region
+func NewClientProvider(authConfig *config.AuthConfig) (*ClientProvider, error) {
+	if err := ociutil.ValidateAuthConfig(authConfig); err != nil {
```

Comment:
the auth config validation will fail in case of instance principal right? if dont think we should validate here, the validation should happen before this in main.go